deps/reftable/include/reftable.h

```diff
@@ -0,0 +1,502 @@
+
+/* returns whether two reftable_ref_records are the same */
+int reftable_ref_record_equal(struct reftable_ref_record *a,
+			      struct reftable_ref_record *b, int hash_size);
+
+/* reftable_log_record holds a reflog entry */
+struct reftable_log_record {
+	char *ref_name;
+	uint64_t update_index; /* logical timestamp of a transactional update.
+				*/
+	uint8_t *new_hash;
+	uint8_t *old_hash;
+	char *name;
+	char *email;
+	uint64_t time;
+	int16_t tz_offset;
+	char *message;
+};
+
+/* returns whether 'ref' represents the deletion of a log record. */
+int reftable_log_record_is_deletion(const struct reftable_log_record *log);
+
+/* frees and nulls all pointer values. */
+void reftable_log_record_clear(struct reftable_log_record *log);
+
+/* returns whether two records are equal. */
+int reftable_log_record_equal(struct reftable_log_record *a,
+			      struct reftable_log_record *b, int hash_size);
+
+/* dumps a reftable_log_record on stdout, for debugging/testing. */
+void reftable_log_record_print(struct reftable_log_record *log, int hash_size);
+
+/****************************************************************
+ Error handling
+
+ Error are signaled with negative integer return values. 0 means success.
+ ****************************************************************/
+
+/* different types of errors */
+enum reftable_error {
+	/* Unexpected file system behavior */
+	IO_ERROR = -2,
+
+	/* Format inconsistency on reading data
+	 */
+	FORMAT_ERROR = -3,
+
+	/* File does not exist. Returned from block_source_from_file(),  because
+	   it needs special handling in stack.
+	*/
+	NOT_EXIST_ERROR = -4,
+
+	/* Trying to write out-of-date data. */
+	LOCK_ERROR = -5,
+
+	/* Misuse of the API:
+	   - on writing a record with NULL ref_name.
+	   - on writing a reftable_ref_record outside the table limits
+	   - on writing a ref or log record before the stack's next_update_index
+	   - on reading a reftable_ref_record from log iterator, or vice versa.
+	*/
+	API_ERROR = -6,
+
+	/* Decompression error */
+	ZLIB_ERROR = -7,
+
+	/* Wrote a table without blocks. */
+	EMPTY_TABLE_ERROR = -8,
+};
+
+/* convert the numeric error code to a string. The string should not be
+ * deallocated. */
+const char *reftable_error_str(int err);
+
+/****************************************************************
+ Writing
+
+ Writing single reftables
+ ****************************************************************/
+
+/* reftable_write_options sets options for writing a single reftable. */
+struct reftable_write_options {
+	/* boolean: do not pad out blocks to block size. */
+	int unpadded;
+
+	/* the blocksize. Should be less than 2^24. */
+	uint32_t block_size;
+
+	/* boolean: do not generate a SHA1 => ref index. */
+	int skip_index_objects;
+
+	/* how often to write complete keys in each block. */
+	int restart_interval;
+
+	/* 4-byte identifier ("sha1", "s256") of the hash.
+	 * Defaults to SHA1 if unset
+	 */
+	uint32_t hash_id;
```

Comment:
I think this doesn't quite work as advertised. I've set up `reftable_write_options opts ={0}`, where I'd expect it to just default to "sha1". But `reftable_new_merged_table()` fails when comparing `r->hash_id != hash_id`.

Reply:
It would also be helpful to have some macros available so that one may set up the expected values without having to define them locally.

Reply:
I can't repro your complaint, see

 https://github.com/google/reftable/commit/8a04e7f2831233bdedaa6a39111200ce4052eff8

can you provide a repro testcase?

Reply:
ping?

Reply:
Sorry, forgot to check in for comments. The following fails if I do not set up `options.hash_id`:

```c
const char *dir, *table;
struct reftable_write_options options = {0};
struct reftable_ref_record ref = {0};
struct reftable_stack *stack;

dir = "/path/to/testrepo-reftable/.git/reftable";
table = "/path/to/testrepo-reftable/.git/reftable/tables.list";

/* This line is required, otherwise `reftable_new_stack` fails. */
options.hash_id = 0x73686131;

cl_must_pass(reftable_new_stack(&stack, dir, table, options));
cl_must_pass(reftable_stack_reload(stack));
cl_must_pass(reftable_stack_read_ref(stack, "refs/heads/master", &ref));
```

Reply:
Note that I'm currently not using an up to date version of the reftable library as I first want to try to get what I have completely functioning (which it mostly is already, but there's still some issues with writing for preexisting branches).

Reply:
which version are you using?

btw, if you are writing existing branches, make sure you sure you set an update_index that is beyond the update_index of the existing branch.

Reply:
thanks, fixed in https://github.com/google/reftable/commit/f781daf143482095be9bba632de61bc0281b8f36